src/librbd/io/ImageDispatchSpec.cc

```diff
@@ -19,6 +19,7 @@ void ImageDispatchSpec::C_Dispatcher::complete(int r) {
       image_dispatch_spec->dispatch_layer - 1);
     [[fallthrough]];
   case DISPATCH_RESULT_CONTINUE:
+  case DISPATCH_RESULT_REFRESH:
```

Comment:
I think this might be your bug - if r < 0 for DISPATCH_RESULT_REFRESH then you've temporarily removed the op from the count, but don't add it back in again before calling fail which will eventually lead to another decrement